drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,588 @@ impl ServerImpl {
         out: &mut [u8],
-    ) -> Result<usize, Error> {
-        let fpga = get_fpga(modules)?;
-        let fpga = self.transceivers.fpga(fpga);
-
-        // This is a bit awkward: the FPGA will get _every_ module's
-        // status (for the given FPGA), then we'll unpack to only the
-        // ones that we care about
-        let enable: U16<LittleEndian> =
-            fpga.read(Addr::QSFP_POWER_EN0).map_err(|_e| {
-                Error::StatusFailed(HwError::PowerEnableReadFailed)
-            })?;
-        let resetl: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_RESETL0)
-            .map_err(|_e| Error::StatusFailed(HwError::ResetLReadFailed))?;
-        let lpmode: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_LPMODE0)
-            .map_err(|_e| Error::StatusFailed(HwError::LpModeReadFailed))?;
-        let presentl: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_MODPRSL0)
-            .map_err(|_e| Error::StatusFailed(HwError::ModPrsLReadFailed))?;
-        let irql: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_MOD_INTL0)
-            .map_err(|_e| Error::StatusFailed(HwError::IntLReadFailed))?;
-        let pg: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_POWER_GOOD0)
-            .map_err(|_e| Error::StatusFailed(HwError::PgReadFailed))?;
-        let pg_to: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_POWER_GOOD_TIMEOUT0)
-            .map_err(|_e| Error::StatusFailed(HwError::PgTimeoutReadFailed))?;
-        let pg_lost: U16<LittleEndian> = fpga
-            .read(Addr::QSFP_POWER_GOOD_LOST0)
-            .map_err(|_e| Error::StatusFailed(HwError::PgLostReadFailed))?;
-
-        // Write one bitfield per active port in the ModuleId
+    ) -> (usize, ModuleId) {
+        let mut error_idx: usize = data_idx;
+        // any modules at index 32->63 are not currently supported.
+        let invalid_modules = ModuleId(0xffffffff00000000);
+        let requested_invalid_modules = ModuleId(modules.0 & invalid_modules.0);
+        // iterate through the results and populate errors in the response buffer
+        // failure: whatever `HwError` specified by `failure_type`
+        // error: fpga communication issue
+        for module in modules.to_indices().map(LogicalPort) {
+            if module <= LogicalPortMask::MAX_PORT_INDEX {
+                if result.failure().is_set(module) {
+                    match failure_type {
+                        Some(e) => {
+                            let err_size =
+                                hubpack::serialize(&mut out[error_idx..], &e)
+                                    .unwrap();
+                            error_idx += err_size;
+                        }
+                        None => {
+                            ringbuf_entry!(Trace::UnexpectedFailure(
+                                ModuleId::from(module.as_mask())
+                            ));
+                            unreachable!();
```

Comment:
Nit: this is still plenty reachable, so I'd recommend a `panic!` instead.

Reply:
Ha, I think I originally recommended `unreachable!()` instead, sorry about that Aaron! I'm actually curious why you'd recommend that @mkeeter.

Reply:
I normally use `unreachable!` if it's **immediately obvious from surrounding context** that you can't get there, but the compiler isn't smart enough to prove it, e.g.
```rust
match i % 4 {
    0 => print!("HI"),
    1 => print!("THERE"),
    2 => print!("omg"),
    3 => print!("wtf"),
    _ => unreachable!(),
}
```

(The [Rust docs](https://doc.rust-lang.org/std/macro.unreachable.html) basically agree with this philosophy)

In this case, it's not immediately obvious from context, so I'd vote for a `panic!`.

Reply:
Sold!